core/trino-main/src/main/java/io/trino/cost/HashPartitionCountProvider.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.cost;
+
+import io.trino.Session;
+import io.trino.SystemSessionProperties;
+import io.trino.operator.RetryPolicy;
+import io.trino.sql.planner.optimizations.PlanNodeSearcher;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.TableScanNode;
+import io.trino.sql.planner.plan.UnnestNode;
+import io.trino.sql.planner.plan.ValuesNode;
+import org.assertj.core.util.VisibleForTesting;
+
+import java.util.List;
+
+import static io.trino.SystemSessionProperties.getFaultTolerantExecutionPartitionCount;
+import static io.trino.SystemSessionProperties.getRemoteHashPartitionMinRowCount;
+import static io.trino.SystemSessionProperties.getRetryPolicy;
+import static io.trino.sql.planner.iterative.Lookup.noLookup;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+import static java.util.Objects.requireNonNull;
+
+public final class HashPartitionCountProvider
```

Comment:
how this relates to `io.trino.cost.TaskCountEstimator`?

If you look we already have `COST_ESTIMATION_WORKER_COUNT` and `io.trino.cost.TaskCountEstimator#estimateSourceDistributedTaskCount`.

There is also `io.trino.cost.TaskCountEstimator#estimateHashedTaskCount`.

We probably should evolve existing `TaskCountEstimator`